README.md
[![Build Status](https://travis-ci.org/vyahello/fake-vehicles-api.svg?branch=master)](https://travis-ci.org/vyahello/fake-vehicles-api)
[![Coverage Status](https://coveralls.io/repos/github/vyahello/fake-vehicles-api/badge.svg?branch=master)](https://coveralls.io/github/vyahello/fake-vehicles-api?branch=master)
[![Issues](https://img.shields.io/github/issues/vyahello/fake-vehicles-api)](https://github.com/vyahello/fake-vehicles-api/issues)
[![EO principles respected here](https://www.elegantobjects.org/badge.svg)](https://www.elegantobjects.org)
[![License](https://img.shields.io/badge/license-MIT-green.svg)](LICENSE.md)
[![Hits-of-Code](https://hitsofcode.com/github/vyahello/fake-vehicles-api)](https://hitsofcode.com/view/github/vyahello/fake-vehicles-api)

# Fake Vehicles API
> A lightweight fake vehicle API written in [API Star](https://docs.apistar.com/) 🌟 python WEB API toolkit framework. 
> This project is built only as a sample REST API toolkit.
>
> Fake data was generated from cool https://www.mockaroo.com source.


## Tools
- `python 3.8`
- `RESTful API`
- `pytest`
- `travis CI`
- `heroku`

In addition it is `fully type annotated` and covered with bunch of static code analysis tools like `mypy`, `flake8`, `pylint`, `pydocstyle` and `black`.

## Usage

![Screenshot](static/demo.png)

### Quick start

Run next command from the root directory of the project:
```bash
python -m api
```

After please open [localhost:5000/](http://localhost:5000/) or [localhost:5000/index.html](http://localhost:5000/index.html) path in your browser to obtain home page. 

## Development notes

### API endpoints
There are several endpoints already developed (others will be added soon):
  - `GET` request on `/api` endpoint to get list of all vehicles e.g:
    ```bash
    curl localhost:5000/api
    ```
  - `POST` request on `/api` endpoint to create a vehicle e.g:
    ```bash
     curl -X POST \
    -H "Content-type: application/json \"
    -d {"manufacturer": "BMW", "model": "X5", "year": 2011, "vin": "WBAWV53589P465532"}' \
    localhost:5000/api
    ```
  - `GET` request on `/api/{id}` endpoint to get single vehicle e.g:
    ```bash
    curl localhost:5000/api/99
    ```
  - `PUT` request on `/api/{id}` endpoint to update single vehicle e.g:
    ```bash
    curl -X PUT \
    -H "Content-type: application/json" \
    -d '{"manufacturer": "BMW", "model": "X5", "year": 2011, "vin": "WBAWV53589P465532"}' \
    localhost:5000/api/99
    ```
  - `DELETE` request on `/api/{id}` to delete single vehicle e.g:
    ```bash
    curl -X DELETE localhost:5000/api/99
    ```

### Testing
Project is covered with both **_unittests_** and **_functional_** tests on REST API endpoints using [pytest](https://docs.pytest.org/en/latest) testing framework.

Please run next command from the root directory to start testing:
```bash
pytest
```
> If you would like to run only `smoke` tests or `unittests` please run tests with corresponding marker e.g:
> 
> ```bash
> pytest -m smoke
> ```

Please open `test-report.html` file in your browser to see testing report.


### Heroku deployment
Please follow instructions from - https://python-responder.org/en/latest/deployment.html

- Install heroku following by - https://devcenter.heroku.com/articles/heroku-cli#download-and-install
- Configure [Procfile](Procfile)
```text
web: gunicorn vehicle_api:api_app
```
- Create `runtime.txt` file
```text
python-3.8.0
```
- Login to heroku
```bash
heroku login
```
- Create an application
```bash
heroku create fake-vehicles-api
```
- Commit and push repo into a heroku
```bash
git add . && git commit -m "My first heroku app" && git push heroku master
```
- Spin up dynos for web
```bash
heroku ps:scale web=1
```
- Check heroku logs
```bash
heroku logs --tail
```
- Open an application via browser: https://fake-vehicles-api.herokuapp.com

## Meta
Author – _Volodymyr Yahello_.

Distributed under the `MIT` license. See [LICENSE](LICENSE.md) for more information.

You can reach out me at:
* [vyahello@gmail.com](vyahello@gmail.com)
* [https://twitter.com/vyahello](https://twitter.com/vyahello)
* [https://www.linkedin.com/in/volodymyr-yahello-821746127](https://www.linkedin.com/in/volodymyr-yahello-821746127)

### Contributing
I would highly appreciate any contribution and support. If you are interested to add your ideas into project please follow next simple steps:

1. Clone the repository
2. Configure `git` for the first time after cloning with your `name` and `email`
3. `pip install -r requirements.txt` to install all project dependencies
4. `pip install -r requirements-dev.txt` to install all development project dependencies
5. Create your feature branch (`git checkout -b feature/fooBar`)
6. Commit your changes (`git commit -am 'Add some fooBar'`)
7. Push to the branch (`git push origin feature/fooBar`)
8. Create a new Pull Request

### What's next

All recent activities and ideas are described at project [issues](https://github.com/vyahello/fake-vehicles-api/issues). 
If you have ideas you want to change/implement please do not hesitate and create an issue.

CHANGELOG.md
* 1.3.1
    * Use "api" endpoint for app APIs
* 1.3.0
    * Deploy app on heroku at https://fake-vehicles-api.herokuapp.com
* 1.2.0
    * Add static code analysis tools
    * Support travis CI
    * Support tests coverage
* 1.1.0
    * Add home page with API description
* 1.0.1
    * Cover app with unittests
* 1.0.0
    * Build basic CRUD API endpoints


api/app.py
from typing import List, Optional
from apistar import App, Route
from apistar.http import JSONResponse
from api import list_of_vehicles, move_to, setup
from api.vehicle.vehicles import Vehicle
from api.web.support import Method, Status, RESPONSE_ERROR


def home() -> str:
    """Return welcome page."""
    return api_app.render_template("home/index.html")


def list_vehicles() -> List[Vehicle]:
    """Returns a list of vehicles.

    Operates "/rest" route on "GET" request e.g "http://127.0.0.1:5000/rest".
    """
    return list(map(lambda vehicle: Vehicle(vehicle[1]), sorted(list_of_vehicles.items())))


def create_vehicle(vehicle: Vehicle) -> JSONResponse:
    """Creates a vehicle.

    Operates "/rest" route on "POST" request e.g "http://127.0.0.1:5000/rest".
    Passes following body structure sample:
    ```
        {
            "manufacturer": "BMW",
            "model": "X5",
            "year": 2011,
            "vin": "WBAWV53589P465532"
        }
    ```

    Args:
        vehicle: a vehicle

    Returns:
        http json response
    """
    vehicle.id_ = len(list_of_vehicles) + 1
    list_of_vehicles[vehicle.id_] = vehicle
    return JSONResponse(vehicle, status_code=Status.CREATED.code)


def get_vehicle(vehicle_id: int) -> JSONResponse:
    """Gets single vehicle.

    Operates "/rest/{id}" route on "GET" request e.g "http://127.0.0.1:5000/rest/99".

    Args:
        vehicle_id: a vehicle ID e.g `2`

    Returns:
        http json response
    """
    vehicle: Optional[Vehicle] = list_of_vehicles.get(vehicle_id)
    if not vehicle:
        return RESPONSE_ERROR
    return JSONResponse(vehicle, status_code=Status.SUCCESS.code)


def update_vehicle(vehicle_id: int, vehicle: Vehicle) -> JSONResponse:
    """Updates vehicle by it's ID.

    Operates "/rest/{id}" route on "PUT" request e.g "http://127.0.0.1:5000/rest/99".

    Passes following body structure sample:
    ```
        {
            "manufacturer": "BMW",
            "model": "X5",
            "year": 2011,
            "vin": "WBAWV53589P465532"
        }
    ```

    Args:
        vehicle_id: vehicle ID e.g `2`
        vehicle: a vehicle item

    Returns:
        http json response
    """
    if not list_of_vehicles.get(vehicle_id):
        return RESPONSE_ERROR
    vehicle.id_ = vehicle_id
    list_of_vehicles[vehicle_id] = vehicle
    return JSONResponse(vehicle, status_code=Status.SUCCESS.code)


def delete_vehicle(vehicle_id: int) -> JSONResponse:
    """Deletes vehicle by it's id.

    Operates "/rest/{id}" route on "DELETE" request e.g "http://127.0.0.1:5000/rest/99"

    Args:
        vehicle_id: vehicle ID

    Returns:
        http json response
    """
    if not list_of_vehicles.get(vehicle_id):
        return RESPONSE_ERROR
    del list_of_vehicles[vehicle_id]
    return JSONResponse({}, status_code=Status.NO_CONTENT.code)


ROUTES = [
    Route(move_to.home, method=Method.GET.name, handler=home),
    Route(move_to.index, method=Method.GET.name, handler=lambda: home()),  # pylint:disable=unnecessary-lambda
    Route(move_to.root, method=Method.GET.name, handler=list_vehicles),
    Route(move_to.root, method=Method.POST.name, handler=create_vehicle),
    Route(move_to.id_, method=Method.GET.name, handler=get_vehicle),
    Route(move_to.id_, method=Method.PUT.name, handler=update_vehicle),
    Route(move_to.id_, method=Method.DELETE.name, handler=delete_vehicle),
]
api_app: App = App(ROUTES, setup.dir_.templates, setup.dir_.static)


if __name__ == "__main__":
    api_app.serve(setup.host, setup.port, setup.debug)
